RFC: Make `any` and `all` short-circuiting

fcard opened the pull request:
<s>Doesn't apply for `n < 16`, but that shouldn't be a big problem, right? There are apparently optimizations for small `n`, so I decided not to mess with that.</s>

fcard committed:
Make `any` and `all` short-circuiting

Deprecate nonboolean `any` and `all`.

Keno commented:
I don't think this is correct. Wouldn't this make all mapreduce invocations short-circuit?

fcard commented:
Only the ones where the reduce function is `AndFun` or `OrFun`. I suppose I could make it more specific.

Keno commented:
Well, at the very least that change deserved additional discussion.

fcard commented:
I see, sorry. It just seemed that this is what the function was supposed to be in the first place, but I guess I shouldn't jump to conclusions so hastily.

Keno commented:
I think we need to determine what the semantics of mapreduce need to be. It is quite possible that the original function is incorrect in short circuiting.

fcard commented:
Yes. If one wants to have side-effects with mapreduce and `|` or `&`, this wouldn't let them. I suppose this is incorrect. It's easy to have `any` and `all` separated from mapreduce, but I wish they could still benefit from it somehow. I could create new functors `AndFunShort` and `OrFunShort` that are specific  to `any` and `all`. I could also try the `isreduced` approach, but from the silence I got I guess nobody is interested :P Maybe short-circuiting is just conceptually incompatible with mapreduce?

fcard commented:
Ok, last try with cleverness, if people don't like this I am just going with the easy route. Either way, after this I am taking a break to reassess my life.

fcard commented:
I realized that the Bool parameter was because `reduce(|, _)` can be used for numbers and that shouldn't short-circuit ever. I could've made a `AbstractArray{Number}` specialization but the more changes I made the more I realized `any`/`all` should've just been its own thing. I also noticed that there are some other functions in _reduce.jl_ that are separated from `mapreduce` and that made me feel more confident in making the simple version.

So here, straightforward version, as it should have been from the beginning. I will see if there are any other changes to be made and then squash. Thanks for your patience.

<sup><sub>darn you lack of impulse control</sub></sup>

sbromberger commented:
@fcard 
> Doesn't apply for n < 16, but that shouldn't be a big problem, right?

This is a big problem. Consider a very expensive function (which is exactly the use case I describe in https://github.com/JuliaLang/julia/issues/11750#issuecomment-113250611). I think the expectation would be "short circuit under ever circumstance" or "short circuit under no circumstance", but not some hybrid that's dependent on the length of the `itr`.

fcard commented:
@sbromberger Fixed it already when I separated it from `mapreduce`. I suppose that's another reason to have done that.

sbromberger commented:
@fcard Ah, sorry. I read your initial comment without having reviewed the change. Thanks!

fcard commented:
@sbromberger No big deal, I realized later that it might have been important to keep it consistent, when `n` is small but the predicate is expensive, for example, or in the case that someone is depending in the short-circuit behavior for side-effects (like counting up to the first `true` or something).

sbromberger commented:
I don't understand why the AppVeyor build is failing. Timeout issues?

fcard commented:
I guess so. From [here](http://www.appveyor.com/pricing):


>Are there any build time restrictions?
>
>All plans have maximum build job execution time of 40 minutes.

tkelman commented:
The worker that started on `linalg1` just froze, ref #7942, happens occasionally. I restarted the build.

sbromberger commented:
w00t - looks like tests passed. Thanks, @fcard - this looks awesome. :)

fcard commented:
I was testing the speed of the new `any` and `all` and noticed that it was slower in a few cases, so I bought back some of the optimizations of old version. Here are the times I got (gists):

[Summarized](https://gist.github.com/fcard/aca7b07d24ff24b01d4d) (average times)
[Expanded](https://gist.github.com/fcard/6c7528bff411696889e9) (max, min, result, averages)

[Here](https://gist.github.com/fcard/f31d0a8d02c7de9f41b1) is the code I used to get the results. It's a bit messy.
I also remove a bit of redundancy from the tests, leftovers from a previous version.

hayd commented on base/reduce.jl:
```diff
@@ -296,46 +296,95 @@ function extrema(itr)
+    i   = 1
+    len = length(itr)
+    while i <= len
+       @inbounds x = itr[i]
```
Is doing this as a while loop faster than `for x in itr` ??

fcard replied:
Apparently. `reduce.jl` is full of tricks like these, and the tests I did seem to confirm this: https://gist.github.com/fcard/aca7b07d24ff24b01d4d#file-sc-any-all-perf-summary-jl-L50-L55

Only for vectors, through, didn't do much for ranges or tuples.

hayd replied:
That's crazy/seems really bad. :( To clarify: the only difference between V1 and V2 in this particular signature is the while/for ?

fcard replied:
Yeah, that's the main difference, the others were made mostly to accommodate it. About it being #11787, maybe? Maybe it's making it worse, but these kinds of optimizations have been on `reduce.jl` for a while, right?

simonster replied:
I'd think `@inbounds for x in itr` should be equally fast. `@inbounds` is necessary for optimal performance due to #11350, but whether we should actually use it for non-Arrays is another question, since we'll segfault on user-defined types if `done` is incorrectly defined.

fcard replied:
I will try changing to this and see if it works faster, although if #11787 is at effect this should obfuscate the results somewhat. Still wondering about the other optimizations in `reduce.jl` though, were they made before the compiler could generate better code for array for loops?

fcard replied:
Yep, `@inbounds` makes it fast. I will keep it in the array definitions only, that segfault problem sounds scary (doesn't seem to make any difference for the other types anyway).

hayd commented on base/reduce.jl:
```diff
@@ -296,46 +296,95 @@ function extrema(itr)
+end
+
+function all(f::IdFun, itr::Array)
+    error("all(itr) expects the elements of itr to be booleans.")
```
Would this be better as an ArgumentError ?

Note: This is an api change. IMO it should be tested as well.

fcard replied:
Actually, maybe I should let these fall back to `reduce(|, itr)` and `reduce(&, itr)`. Both `all` and `any` still accept BitArrays after all.

Then again, it would not be consistent, as this wouldn't be short-circuiting.

Maybe it should be discussed what to do about "bitwise any/all". Deprecate it? Let it stay?

ScottPJones replied:
I'm not sure I understand, you want to not allow `any()` or `all()` to work (and short-circuit) on BitArrays?  That would be sad...  that's the sort of thing you want for SQL queries, with a bitmap index, right?

hayd replied:
I think so. IMO deprecate is the thing to do, `reduce(|, itr)` is much clearer than `any(itr)`.

ScottPJones replied:
And `reduce(|, itr)` short-circuits, or not?  (sorry for the dumb question!)

fcard replied:
@ScottPJones  Oh no, BitArrays should stay in my opinion, if you look at the performance tests, they are the fastest of them all :P And they have been always short-circuiting, so an example to follow, not to reprimand.

Since any used to just be pretty much a alias for `reduce(|, itr)`, they both short-circuited on the same cases. It won't anymore, to let people depend on the property of reduce to always go through all elements. Not a dumb question, it's pretty confusing, which is what we're trying to fix here :)

@hayd I think so too, and `any/all` returning numbers always seemed weird to me. Should I will go ask for opnions on the issue?

sbromberger replied:
@fcard my opinion, for what it's worth: `any` / `all` should return Bool. That's what's specified in the documentation*. Either we change the documentation to specify this additional behavior, or we make it strict.

*
```
Base.any(itr) -> Bool

   Test whether any elements of a boolean collection are true.

Base.any(A, dims)

   Test whether any values along the given dimensions of an array are
   true.

Base.any(p, itr) -> Bool

   Determine whether predicate "p" returns true for any elements of
   "itr".

Base.all(itr) -> Bool

   Test whether all elements of a boolean collection are true.

Base.all(A, dims)

   Test whether all values along the given dimensions of an array are
   true.

Base.all(p, itr) -> Bool

   Determine whether predicate "p" returns true for all elements of
   "itr".

      julia> all(i->(4<=i<=6), [4,5,6])
      true
```

fcard replied:
@sbromberger Agreed. For now I am going to put a fallback with a comment, while I wait to see if anybody disagrees.

simonster commented on base/reduce.jl:
```diff
@@ -296,46 +296,95 @@ function extrema(itr)
     end
-    return true
+    return false
 end
```
Are you sure you need this method? The method below should generate identical code if `f` is `IdFun`.

fcard replied:
Oh yeah, that's right. Although I am not even sure I can use IdFun because `identity` is defined after any is first used (cause of latest build failure). Is there problem in moving `identity`'s  definition earlier?

simonster commented:
IMO `mapreduce` should short-circuit but `mapfoldl`/`mapfoldr` should not. `mapreduce` already leaves many things up to the implementation. We don't define the associativity and don't guarantee that the map function is called for each element of the array if they are non-unique to avoid having to invoke `f` on all the zeros in a sparse matrix. Perhaps we should just say that `mapreduce` assumes that `f` is "pure" modulo allocation, i.e., that invoking `f` on the same input produces outputs that are `==` and of the same type, and invoking `f` does not modify observable global state.

fcard commented:
I addressed the feedback I received so far. There are some things being debated, so I am going keep this as a WIP for now and wait until conclusions are made.

Another thing that I need to do is solve the problem I reached when trying to optimize `any`/`all` when using the `identity` function. It's not really necessary for this, so if it becomes too much trouble I will leave it for later, but it's just that I find it unintuitive that `any(identity, itr)` is 10x slower than `any(itr)`.

nalimilan commented:
+1 to @simonster's proposal. Also, I agree `any` and `all` should only return `Bool`.

ScottPJones commented:
Ditto

fcard commented:
Hello, I am mostly done implementing this, but I reached a stumbling block.
Is this a valid usage of `mapreduce`? 

```julia
julia> mapreduce(x->(x == 1? true : x), |, 1:10)
15
```

Or using `f::T -> Union{Bool, SomethingElse}` in general. If we reach an `true`, how do we know if there are any nonboolean results later that might modify the final result? In the case of `any` or `all`, it's a safe assumption to make that the predicate should return a `Bool`, but with `mapreduce`, I'd expect the transformer function to be allowed to return any value that can be used with the operator, and `|`/`&` can use booleans, integers and arrays, all in the same operation.

(The current implementation actually assumes that predicates return booleans so this happens:
```julia
julia> mapreduce(Int, |, trues(15))
1

julia> mapreduce(Int, |, trues(16))
ERROR: TypeError: non-boolean (Int64) used in boolean context
 in mapreduce_impl at reduce.jl:327
 in _mapreduce at reduce.jl:151
 in mapreduce at reduce.jl:158
```
)

nalimilan commented:
Calling `|` or `&` on a mix of booleans and other types doesn't make any sense to me. The fact that these operators have quite different meanings strikes again... Cf. https://github.com/JuliaLang/julia/issues/5187

So I'd say the question is: is it OK to fail that way, or should a better error be reported? Maybe that's not an issue as long as the docs clearly state that with `|` and `&`, `mapreduce` will short-circuit as soon as some values are encountered (giving a clear definition of them).

fcard commented:
Iunno, there are more than 30 methods in both `|` and `&` and half of them involve two potentially different types. In my opinion, if these mixups are unwanted, the methods themselves should be removed rather than the support dropped in mapreduce. I mean, unless there are no other solutions.

The way I found to deal with this is the following:

-
In `mapreduce(f, op, A::Iterable{T})`, infer what are the possible return types of `f(T)`. If the only return type is a type of which we have a single value to short-circuit from with `op`, then `mapreduce` short-circuits on that value.

For example, in `mapreduce(f=identity, op=|, A=[false, true, false])`, A is an iterable of `Bool`, and `identity(Bool) -> Bool`, and since we can exit on `true` with `|`, then we short-circuit as soon as we reach the second element of `A`.

In the case of `f(x) = x == 1? true : x`, `mapreduce(f, |, 1:10)` can't short-circuit because there is no single value to exit from since `f(Int) -> Union{Int, Bool}`.

Both `any` and `all` create a special `Predicate` functor that always assumes it's return value is `Bool`, so these two functions always try to short-circuit. (and error out if the predicate returns a non-boolean type)

-

Unfortunately this made the code much more complex than I wish it were (although surprising not slower in most cases, in fact it's faster in some), so alternate solutions are welcome.
(It's also not done yet, I am still deciding on a few things)

simonster commented:
I agree that  `mapreduce` has to give the same results as the equivalent loop for `|` and `&`. As annoying as it is, anything else seems difficult to specify and reason about. What you describe above seems like a reasonable solution, although it is complex. I guess I'm not sure whether mapreduce will be used often enough with `|` and `&` in cases where we can short-circuit that it's worth worrying about optimizing. Maybe better to avoid type inference and just optimize `Predicate`?

fcard commented:
I have removed the runtime type inference, so currently only `any([f, ]itr)`, `all([f, ]itr)` and `reduce(|/&, itr::AbstractArray{Bool})`<sup>[ ]</sup>  are short-circuiting. There is nothing else that I want to do,
any comments/requests?

-
<sub>[ ] making the `reduce` case work on all iterables would involve some changes on possibly performance sensitive areas, and since that's outside of this PR's scope I am going to leave it as it is for now to keep things simple.</sub>

ScottPJones commented:
They probably will want to squash at least some of the commits, nice stuff!

tkelman commented:
There are many many warnings here during the tests due to the newly added deprecations.

fcard commented:
@ScottPJones I tend to wait until I know there is nothing to change before I squash everything. And thanks :)

@tkelman Just saw it. Is this deprecation completely accepted? If so, I will go ahead and change the instances of `all` to the appropriate methods.

ScottPJones commented:
Well, until the new deprecations, I thought it was nothing to change! :grinning:  Trying to keep up with the rapid pace of master can be tiring!

fcard commented:
It seems that the problematic cases are instances of `all/any(map(f, A))` or `all/any([f(x) for x in A])`, and I guess it's good that these were caught because using map/comprehensions inside `any/all` impedes short-circuiting, they should've been `all/any(f, A)` instead. Maybe I should put a warning in the deprecation about these uses.

fcard committed:
Remove maps/comprehensions from uses of any/all

fcard commented:
There you go, all clean.

tkelman commented:
Looks like there's a legitimate problem on 32 bit
```
exception on 2: ERROR: LoadError: test failed: !(isequal(s.dict.keys,k))
 in expression: !(isequal(s.dict.keys,k))
 in error at error.jl:21
 in default_handler at test.jl:29
 in do_test at test.jl:52
 in include at boot.jl:254
 in runtests at C:\projects\julia\test\testdefs.jl:197
 in anonymous at multi.jl:837
 in run_work_thunk at multi.jl:590
 in anonymous at task.jl:837
while loading C:\projects\julia\test\sets.jl, in expression starting on line 95
	From worker 3:	     * replutil             in  10.43 seconds
ERROR: LoadError: LoadError: test failed: !(isequal(s.dict.keys,k))
 in expression: !(isequal(s.dict.keys,k))
 in anonymous at task.jl:1403
while loading C:\projects\julia\test\sets.jl, in expression starting on line 95
while loading C:\projects\julia\test\runtests.jl, in expression starting on line 5
 
	From worker 2:	     * sets                Command exited with code 1
```

fcard commented:
No prob. I got that same `!(isequal(s.dict.keys,k))` error twice on a previous build, so maybe it's unrelated? It seems to be about rehashing, any recent changes to that?

And is there something I can do about that build failure? I can't seem to find the appointed file to be able to investigate it.

sbromberger commented:
@fcard is this effort abandoned?

fcard commented:
Please just forget about me. I am too incompetent to help, and you people don't have time to correct my mistakes. I am sick of being a burden, I don't want to feel like this anymore, just let me be.

sbromberger commented:
I don't understand. Are we (because of your efforts) not close to a solution here? I'm counting on short-circuit and am likely the least-qualified person in this thread to provide working code. :(

hayd commented:
@fcard Please don't be disheartened if things take a while to merge, you're no burden. This is good and very useful work! (The appveyor failure is unrelated IIUC.)

ScottPJones commented:
@fcard - take it from somebody who has had a lot of sharp review here, you are not incompetent, nor a burden.  People were crazy busy with JuliaCon, and then a lot of people celebrating the 4th of July here in the States, so things slowed down a bit.  (Also, I'm still trying to get a change through after 2.5 months,
so I know it is easy to get impatient, but everybody is donating their time, so sometimes things take a while)

tkelman commented:
@fcard we're terribly sorry this seems to have fallen off the radar. Your contributions are absolutely welcome, we just seem to have gotten briefly stuck and missed getting back to them. You can leave a reminder message once every few days to ask for feedback and serve as a reminder that there's an open PR with contributions you'd like to make, and we can work to get bug fixes like this one merged. It may take a while and some persistence to get through review since everyone's busy.

fcard commented:
Sorry, I had a little mental breakdown. Thank you very much for your messages, they made me feel a lot better. I am really not in the mental state to actively contribute to anything right now (haven't been for a while to be honest), but I do want to at least finish this one before I take a break, it would give me some peace at least.

@tkelman it this is anybody's fault, it's mine, so no need to apologize. I will persist.

StefanKarpinski commented:
@fcard, just so you know, your effort here really is very much appreciated.

ViralBShah commented:
Yes, second that.

Mike43110 commented:
@fcard Good luck! This is a welcomed change.
I do hope you feel better, this is quite a random comment considering we are on GitHub!
That being said: Get your head straight however you need to. It is a scary place to be in - sometimes the problems are real sometimes they are all just in your head. All that matters is that they are real to you.
Once again, good luck. I hope that you feel better soon.

pao commented:
Can we reiterate the action items here? Or is this a rebase-and-kick-AppVeyor to see where we stand kind of thing? https://github.com/JuliaLang/julia/pull/11774#issuecomment-116010218 looks like the only open item but I may be mistaken.

tkelman commented:
A rebase would be a good first step, the code looks good to me and I suspect the appveyor failure was likely caused by something else

tkelman commented:
Oh dear, there's a freeze while building inference on win64, these can be really hard to debug unfortunately. I can reproduce it locally, though I'm about to get on a plane.

yuyichao commented:
@tkelman Just curious, have you been able to reproduce any of these windows freezing issues with LLVM 3.6?

tkelman commented:
@yuyichao I haven't tried this particular one with LLVM 3.6 yet, but in a few past cases of similar issues  LLVM 3.6 made codegen freezes like this one go away.

Anyone have any code tweaking suggestions here that might make LLVM 3.3 less finicky on Win64? @mbauman you had hit a bunch of these, can you offer any guidance?

mbauman commented:
No magic bullets, I'm afraid. As I recall, I just went back to a known working state and slowly "bisected" my patchset. The other solution was to find usages in bootstrap and remove them.

yuyichao commented:
@tkelman The reason I asked is that your last debugging of the freeze points to memory managing code of the old JIT (or the old JIT in general). Even if it's not a LLVM bug, it'll probably be very different with the MCJIT.

tkelman commented:
Indeed, LLVM 3.6.1 does not hit the bootstrap freeze.

@fcard just so you're clear, this problem is in no way your fault. It just so happens that `any` and `all` get used by type inference, which is a core part of the Julia compiler so changes to it are more likely to trigger bugs, especially with LLVM 3.3 and win64. If you're not able to try experimenting with this code, trying a few different things out (or waiting until we upgrade LLVM but that may take a while ref #9336) until we find something that passes on all platforms, just let us know.

One thing we could possibly change is to modify inference.jl so it doesn't use the short-circuiting versions of `any` or `all` for now.

fcard commented:
I don't mind doing some more work. I will see what I can do about this.

tkelman commented:
You might not have to do anything after all. I restarted the build and it looks like this comes and goes depending on the state of master. So maybe it's safe for now, but we keep skirting the edge of this bug (whatever is causing it) based on totally unrelated changes.

tkelman commented:
Looks like @yuyichao fixed the underlying issue, hooray!

So what do people think, should we merge this now?

phobon commented:
Is there any case to make that callable objects should also be allowed? I.e., are all the type asserts of `Function` required?

Technically this is a different issue because it doesn't currently work on master either. Just something to think about while the behavior is being changed anyway.

tkelman commented:
sounds a bit like #12035

timholy commented:
:+1: from me. Unfortunately it needs a rebase. Once done, I would favor merging this.

lindahua commented on base/reduce.jl:
```diff
@@ -155,12 +155,60 @@ end
+mapreduce_no_sc(f, op, itr::Any)           =  mapfoldl(f, op, itr)
+mapreduce_no_sc(f, op, itr::AbstractArray) = _mapreduce(f, op, itr)
+
+mapreduce_sc(f::Function,    op, itr) = mapreduce_sc(specialized_unary(f), op, itr)
```
You might want to add a comment to document what `mapreduce_no_sc` and `mapreduce_sc` mean. When I saw the abbreviation `sc`, I was not able to immediately associate it with *short circuit*.

fcard replied:
@lindahua If I change the `## mapreduce definitions` to `## short-circuiting (sc) mapreduce definitions`, will it be clear enough?

lindahua replied:
Looks clear enough. Thanks.

lindahua commented:
Here is my concern. It looks like that the behavior depends on the type of `f` (*i.e.* whether it is to a predicate or not, etc). This precludes the short-circuit behavior to be applied when `f` is a customized callable instance.

What about introducing functions like `all_sc` and `any_sc`, which allows opt-in at the calling site?

fcard commented:
Rebased!

@lindahua What you're saying applies to `mapreduce`, but there shouldn't be any problem making `any` and `all` work on all callables, since they would get wrapped in a `Predicate` anyway. I don't mind making this change, if people want it.

About having separate functions, this was discussed at the beginning of #11750, but the conclusion I think was that people shouldn't be relying on non-short-circuiting behaviour for side-effects while using predicates. One could still use `mapfoldl(f, |\&, A)` if they don't want short-circuiting.

lindahua commented on base/functors.jl:
```diff
@@ -61,6 +61,20 @@ end
+
+type Predicate <: Func{1}
+    f::Function
+end
```
With such definition, we are not able to encapsulate a callable *functor* (not *function*) into an instance of `Predicate`.

For example:
```julia
immutable IsPosFun end
call(::IsPosFun, x) = (x > zero(x))

# The following statement would cause an error
Predicate(IsPosFun())
```

Use of functors is quite important, especially in cases where performance is critical.

lindahua commented:
Here is an example:

```julia
immutable IsPosFun end
call(::IsPosFun, x) = (x > zero(x))

# people might want to have short-circuit behavior on the following
# but with current implementation, it does not seem to be the case
all(IsPosFun(), x)  

# also, `Predicate(IsPosFun())` does not work
```

lindahua commented:
I think the `Predicate` type should be defined in a way to allow it to encapsulate a functor, as follows:

```julia
immutable Predicate{F}
    f::F
end
Predicate{F}(f::F) = Predicate{F}(f)

call(pred::Predicate, x) = call(pred.f, x)
```

fcard commented:
Yeah, that's what I thought I had to do, make both the predicate and `any`/`all` receive objects other than functions. It's a pretty simple to do, I am just waiting to see if everyone agrees with the change.

tkelman commented:
+1 to @lindahua's suggestion

fcard committed:
Make any/all capable of receiving functors.

fcard commented:
Done. Since this commit is not related to short circuiting, should I keep separate, or should I squash it regardless?

tkelman commented:
I think separate commits are fine for this change, as long as the tests would pass if someone were to happen upon one of the intermediate commits here in the process of doing `git bisect` on something else.

sbromberger commented:
I'm coming back to this after a while away, but I confess to being completely lost. It seems that we might have diverged a bit from my original "make `any()` and `all()` short-circuiting" request and I'd just like to confirm that it's implemented in this PR. :)

Thanks,

Seth.

fcard commented:
@tkelman The tests pass on all commits, but IIRC the first one triggered a lot of deprecations; would that cause any trouble?

@sbromberger That's still there, don't worry. This functor thing is a "might as well do this too while we're here" kinda change, short-circuiting remains the main objective.

tkelman commented:
You've rebased this more than enough times by now. Thanks for all the hard work here! :confetti_ball:

fcard commented:
Thank you very much!  I wouldn't mind doing some more work, but I am happy to see this merged. Hopefully it will be useful :)

Thanks again everybody for the kind messages, I didn't mention the new ones before because I didn't want to derail this further, but I am really grateful. Thanks also for your time and patience.

If there is a next time, let's hope it will go a lot smoother. Till then!

sbromberger commented:
This. Is. Awesome. Thank you all, especially @fcard, for getting this changed.

tkelman commented:
I certainly hope there will be a next time, this was really well done. But only if you're up for it, of course.

fcard commented:
I am! And I am feeling a lot better, but I want to make sure I am 100% before I make another contribution. I will give myself a cool down, work on a few personal projects, and then I will resume, starting with that deprecations testing I said I would do.

Thanks again, contributing to Julia was a very positive experience. I learned a lot, and the community was very welcoming. I hope to return soon :)

StefanKarpinski commented:
We look forward to it – thanks for contributing!
